cmake/HPX_SetupHPXLocal.cmake

```diff
@@ -15,25 +15,120 @@ elseif(NOT TARGET HPX::hpx_local AND NOT HPX_FIND_PACKAGE)
+      EXAMPLES_OPENMP
+      EXAMPLES_QTHREADS
+      EXAMPLES_TBB
+      GENERIC_CONTEXT_COROUTINES
       HIP
       ITTNOTIFY
       MALLOC
       NICE_THREADLEVEL
-      PARCELPORT_MPI_ENV
-      PARCELPORT_MPI_MULTITHREADED
       SANITIZERS
       SPINLOCK_DEADLOCK_DETECTION
       VALGRIND
   )
   foreach(compat_option ${HPX_HPXLOCAL_COMPATIBILITY_OPTIONS})
     if(DEFINED HPX_WITH_${compat_option})
-      hpx_error(
-        "HPX_WITH_${compat_option} is set (\"${HPX_WITH_${compat_option}}\"), but HPXLocal_${compat_option} should be used instead. Forwarding HPX_WITH_${compat_option} to HPXLocal_WITH_${compat_option} for compatibility, but HPX_WITH_${compat_option} will be ignored in the future."
-      )
-      set(HPXLocal_WITH_${compat_option}
-          "${HPX_WITH_${compat_option}}"
-          CACHE STRING "" FORCE
-      )
+      if(NOT DEFINED HPXLocal_WITH_${compat_option})
+        hpx_warn(
+          "HPX_WITH_${compat_option} is set (\"${HPX_WITH_${compat_option}}\"), but HPXLocal_WITH_${compat_option} should be used instead. Forwarding HPX_WITH_${compat_option} to HPXLocal_WITH_${compat_option} for compatibility, but HPX_WITH_${compat_option} will be ignored in the future."
+        )
+
+        get_property(
+          option_type
+          CACHE "HPX_WITH_${compat_option}"
+          PROPERTY TYPE
+        )
+        if(NOT option_type)
+          set(option_type BOOL)
+        endif()
+
+        get_property(
+          option_description
+          CACHE "HPX_WITH_${config_option}"
+          PROPERTY HELPSTRING
+        )
+        if(NOT option_type)
+          set(option_description "")
+        endif()
+
+        set(HPXLocal_WITH_${compat_option}
+            "${HPX_WITH_${compat_option}}"
+            CACHE ${option_type} ${option_description} FORCE
+        )
+      else()
+        hpx_info(
+          "HPX_WITH_${compat_option} is set (\"${HPX_WITH_${compat_option}}\"), but HPXLocal_WITH_${compat_option} is set as well (\"${HPXLocal_WITH_${compat_option}}\"). Ignoring HPX_WITH_${compat_option}."
+        )
+      endif()
+    endif()
+  endforeach()
+
+  # Propagate options to HPXLocal, if needed
+  set(HPX_HPXLOCAL_CONFIGURATION_OPTIONS
+      APEX
+      ATTACH_DEBUGGER_ON_TEST_FAILURE
+      CHECK_MODULE_DEPENDENCIES
+      COMPILE_ONLY_TESTS
+      COMPILER_WARNINGS
+      COMPILER_WARNINGS_AS_ERRORS
+      COROUTINE_COUNTERS
+      DEPRECATION_WARNINGS
+      DISABLED_SIGNAL_EXCEPTION_HANDLERS
+      EXAMPLES
+      EXECUTABLE_PREFIX
+      FAIL_COMPILE_TESTS
+      FETCH_ASIO
+      FULL_RPATH
+      GCC_VERSION_CHECK
+      HIDDEN_VISIBILITY
+      LOGGING
+      PARCELPORT_MPI_ENV
+      PARCELPORT_MPI_MULTITHREADED
+      TESTS
+      TESTS_BENCHMARKS
+      TESTS_EXAMPLES
+      TESTS_EXTERNAL_BUILD
+      TESTS_HEADERS
+      TESTS_REGRESSIONS
+      TESTS_UNIT
+      TOOLS
+      UNITY_BUILD
+      VERIFY_LOCKS
+      VERIFY_LOCKS_BACKTRACE
+      VERIFY_LOCKS_GLOBALLY
+      VS_STARTUP_PROJECT
+  )
```

Comment:
Could you explain what the intended difference between the two lists of variables here are?

My intention was that most of the build-system related options are completely independent and should not be forwarded, but then you will be using this so I'll go along with whatever you find most convenient.

Reply:
The first list handles options that are now supported by HPXLocal only and users should change their build scripts to adjust. Those options have no meaning for the full HPX anymore and would be ignored there in any case. 

The second list of options are still significant for the full HPX and in order to maintain backwards compatibility for its users I think they should be propagated directly (i.e. if `HPX_WITH_TESTS` is enabled, the `HPXLocal_WITH_TESTS` should be too, etc.). For me one of the goals at this point is to make the transition for users of the full HPX as seamless as possible, for them in the end everything should stay unchanged.

Reply:
Ok, then I think the following are definitely ok in this list (build system-related and there are options for both local and distributed):
- CHECK_MODULE_DEPENDENCIES
- COMPILE_ONLY_TESTS
- COMPILER_WARNINGS
- COMPILER_WARNINGS_AS_ERRORS
- DEPRECATION_WARNINGS
- EXAMPLES
- EXECUTABLE_PREFIX
- FAIL_COMPILE_TESTS
- GCC_VERSION_CHECK
- TESTS
- TESTS_BENCHMARKS
- TESTS_EXAMPLES
- TESTS_EXTERNAL_BUILD
- TESTS_HEADERS
- TESTS_REGRESSIONS
- TESTS_UNIT
- TOOLS
- UNITY_BUILD
- VS_STARTUP_PROJECT

The following only have local options and I think they should be in the first list:
- FETCH_ASIO: Because of the reasons mentioned in other comments.
- VERIFY_LOCKS: This is local-only.
- VERIFY_LOCKS_BACKTRACE: This goes with the previous option.

The following currently only have local options. If they are to stay in this list the corresponding distributed option should stay.
- APEX: HPX_WITH_APEX=ON would require HPXLocal_WITH_APEX=ON, so I initially only left the latter. It probably makes more sense to have both though.
- ATTACH_DEBUGGER_ON_TEST_FAILURE: A HPX_WITH_ATTACH_DEBUGGER_ON_TEST_FAILURE would have no effect with the current runtime configuration. This is only controlled by the local option.
- COROUTINE_COUNTERS: HPX_WITH_COROUTINE_COUNTERS=ON would require HPXLocal_WITH_COROUTINE_COUNTERS=ON, which is why I only left the latter like for APEX. Having both makes sense though.
- DISABLED_SIGNAL_EXCEPTION_HANDLERS: This is enabled/disabled when starting the runtime, so having both options makes sense.
- LOGGING: This is enabled/disabled when starting the runtime, so having both options makes sense. HPX_WITH_LOGGING=ON requires HPXLocal_WITH_LOGGING=ON, however.
- FULL_RPATH: Independent, so can have both.
- HIDDEN_VISIBILITY: Only local because it means the export definitions are the same. They could be separate though.
- PARCELPORT_MPI_ENV/PARCELPORT_MPI_MULTITHREADED: These are a bit unfortunate. The local version creates the define that is used by both local and distributed. The distributed one has no effect (it's not an option right now). Should they be separate?
- VERIFY_LOCKS_GLOBALLY: This is currently all in the local repo, but it looks like it would be straightforward to move e.g. to the distributed `init_runtime` module or its own module; should I go ahead and do that?

Let me know if you have preferences on any of these.

Reply:
A few corrections:
- `COROUTINE_COUNTERS`: I don't think it makes sense to have a separate variable for distributed here. If HPXLocal has it enabled one is already paying for the runtime cost of keeping track of those counters. I don't see a use case where it would make sense to give the two variables different values. 
- `FULL_RPATH`: This already had separate variables for local and distributed. Nothing to do here.

Reply:
Mikael, thanks for the thorough analysis. I mostly agree with what you said. A few comments, though:

- `COROUTINE_COUNTERS`: this enables two things: the counting itself in local and the corresponding performance counters in full. Having the latter without the former will break things. I think it should be propagated.
- `VERIFY_LOCKS_GLOBALLY`: we should keep all the locking checks in one place, so moving everything into its own module might be a good idea.
- `FETCH_ASIO`: I agree that in principle this belongs into the first list only. However, this is one of the options people have added just recently to their build scripts. I wanted to avoid breaking those, thus I added it to the second list. My rational is that FETCH_ASIO is one of the important build variables probably used by many people, changing it might be too disruptive.
- `PARCELPORT_MPI_ENV`: isn't this used by the parcelport?

As said, in a perfect world nothing would change for full HPX users. This is especially important as the same build scripts will be used for different versions of HPX on different machines or different configurations. So, I even think that the handling of variables in the first list should be changed at this point enabling that compatibility, while issuing warnings that things will change in the future.

Reply:
> - `COROUTINE_COUNTERS`: this enables two things: the counting itself in local and the corresponding performance counters in full. Having the latter without the former will break things. I think it should be propagated.

Yes. See my correction. But it should be propagated through the first list of variables. If there's only one variable there's no chance of them being out of sync.

> - `VERIFY_LOCKS_GLOBALLY`: we should keep all the locking checks in one place, so moving everything into its own module might be a good idea.

See https://github.com/STEllAR-GROUP/hpx/pull/5684 by the way. I've moved it to `init_runtime` there but a separate module is also fine.

> - `FETCH_ASIO`: I agree that in principle this belongs into the first list only. However, this is one of the options people have added just recently to their build scripts. I wanted to avoid breaking those, thus I added it to the second list. My rational is that FETCH_ASIO is one of the important build variables probably used by many people, changing it might be too disruptive.

But so are e.g. `HPX_WITH_MALLOC` and `HPX_WITH_CUDA`. Why would they be in the first list, but `HPX_WITH_FETCH_ASIO` in the second? I still think `HPX_WITH_FETCH_ASIO` belongs in the first list.

> - `PARCELPORT_MPI_ENV`: isn't this used by the parcelport?

Yes, the preprocessor definition. But that is propagated from the local config. It's also used for detecting the MPI environment in the local project. The name is very unfortunate for this double use case. We can have them separate and rename the ones in the local project to something more descriptive.

> As said, in a perfect world nothing would change for full HPX users. This is especially important as the same build scripts will be used for different versions of HPX on different machines or different configurations. So, I even think that the handling of variables in the first list should be changed at this point enabling that compatibility, while issuing warnings that things will change in the future.

Understood. I'm only concerned about which list these variables end up on, not whether they should be forwarded at all. Variables in the first list still just give a warning, so things will keep working as before even if variables are in that list.

Reply:
I've updated this PR based on the discussion here. I'll go ahead and merge it but don't see this as set in stone. Let's change things afterwards where it makes sense. I still think at least that the `PARCELPORT_MPI_X` variables may need new names on the local side...